### PR TITLE
Packet Coalescing - Core

### PR DESCRIPTION
**Pull request by muxi**
Add packet coalescing feature to gRPC on iOS. Ops in the same batch will be sent within as few packets as possible.

Sub-PR of #9010. This PR consists of changes in core.

### CODE REVIEW

File: src/core/ext/transport/cronet/transport/cronet_transport.c [hunk 1]
```diff
@@ -151,6 +151,11 @@ struct op_state {
   bool state_callback_received[OP_NUM_OPS];
   bool fail_state;
   bool flush_read;
+#ifdef GRPC_CRONET_WITH_PACKET_COALESCING
```
Comment (ctiller) [inline src/core/ext/transport/cronet/transport/cronet_transport.c]: In general it's better to avoid this style of feature macro: only one branch ever gets really tested, and logic gets more and more convoluted around it.

File: test/core/end2end/tests/packet_coalescing.c [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+  shutdown_client(f);
+
+  grpc_completion_queue_shutdown(f->cq);
+  drain_cq(f->cq);
+  grpc_completion_queue_destroy(f->cq);
+}
+
+static bool coalesced_message_and_eos;
+
+static void log_processor(gpr_log_func_args *args) {
```
Comment (ctiller) [inline test/core/end2end/tests/packet_coalescing.c]: This is really objectionable: hardcoding filenames and log formats into a test makes this a brittle change detector.

Please find some other way of verifying this prior to submission.

File: third_party/objective_c/Cronet/cronet_c_for_grpc.h [hunk 3]
```diff
@@ -5,6 +5,8 @@
 #ifndef COMPONENTS_CRONET_IOS_CRONET_C_FOR_GRPC_H_
 #define COMPONENTS_CRONET_IOS_CRONET_C_FOR_GRPC_H_
 
+#define CRONET_EXPORT __attribute__((visibility("default")))
```
Comment (ctiller) [inline third_party/objective_c/Cronet/cronet_c_for_grpc.h]: Why is this needed?

File: test/core/end2end/tests/packet_coalescing.c [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+  grpc_byte_buffer_destroy(request_payload);
+  grpc_byte_buffer_destroy(request_payload_recv);
+
+  end_test(&f);
+  config.tear_down_data(&f);
+}
+
+void packet_coalescing(grpc_end2end_test_config config) {
+  /* The test case does not support the fixture socketpair_one_byte_at_a_time */
```
Comment (ctiller) [inline test/core/end2end/tests/packet_coalescing.c]: Better to use a feature mask instead of a brittle name check.

File: test/core/end2end/tests/packet_coalescing.c (hunk 2 above)
Comment (muxi) [inline test/core/end2end/tests/packet_coalescing.c]: @ctiller Can we have some wisdom on how to test this? Testing against log is indeed awkward, but we could hardly come up with another way as whether frames are coalesced into one packet or not appears transparent to the test case driver.

File: third_party/objective_c/Cronet/cronet_c_for_grpc.h (hunk 3 above)
Comment (muxi) [inline third_party/objective_c/Cronet/cronet_c_for_grpc.h]: This entire file is copied from the iOS framework provided by Cronet.

File: test/core/end2end/tests/packet_coalescing.c (hunk 2 above)
Comment (ctiller) [inline test/core/end2end/tests/packet_coalescing.c]: I'm thinking something like the tcp endpoint overrides that @murgatroid99 has proposed in the past: the test could override the endpoint creation with a wrapper that counts the number of writes that occur on the endpoint.

If we take this test out of the end2end suite (which I think is fine), we could implement such an endpoint, wrap it in a channel, and test it directly without the extra machinery.

Comment (muxi) [inline test/core/end2end/tests/packet_coalescing.c]: A problem is that the test will be used (actually mainly be used) for Cronet, and the endpoint of Cronet is inside the Cronet library which cannot be overriden.

FWIW, following your suggestion I am thinking of either of the following ways:
1. Create a server like what e2e test does and override the reading endpoint of the server
2. Override entire Cronet layer (requires making two different test case drivers for Cronet and chttp2)
3. Make the test for Cronet only. chttp2 and Cronet behavior will be potentially different.
Do you think any of them reasonable?

**Comment by ctiller** [comment]
I'd lean towards (3) here.

File: test/core/end2end/tests/packet_coalescing.c (hunk 2 above)
Comment (ctiller) [inline test/core/end2end/tests/packet_coalescing.c]: Now we just need to find a way to test the property we want without scraping logs.

Comment (muxi) [inline test/core/end2end/tests/packet_coalescing.c]: Right, and I am afraid we still need either (1) or (2) for it as we have no control inside Cronet

Comment (muxi) [inline test/core/end2end/tests/packet_coalescing.c]: Maybe another option is to add some testing code and counter in the core and export them for testing?

**Comment by ctiller** [comment]
Or to write an independent server that verifies the properties you want?
Something akin to the http2 server Makarand has been putting together?

File: test/core/end2end/tests/packet_coalescing.c (hunk 2 above)
Comment (jcanizales) [inline test/core/end2end/tests/packet_coalescing.c]: What is the desired property we're achieving with the coalescing? Latency? Bandwidth? Can we exaggerate it with the test, so that the distinction between coalescing or not becomes readily apparent?

Comment (muxi) [inline test/core/end2end/tests/packet_coalescing.c]: Mostly for latency. I think @makdharma talked about a doc with discussion on this. We planed to test latency with Fireball after they deploy it.

File: src/core/ext/transport/cronet/transport/cronet_transport.c (hunk 1 above)
Comment (muxi) [inline src/core/ext/transport/cronet/transport/cronet_transport.c]: @makdharma is current framework allow any configuration information (e.g. enable_packet_coalescing) into cronet_transport? Using initial metadata seems to be a way to me but I am wondering if there is any dedicated approach.

File: test/core/end2end/tests/packet_coalescing.c (hunk 2 above)
Comment (jcanizales) [inline test/core/end2end/tests/packet_coalescing.c]: > We planed to test latency with Allo after they deploy it.

The super ideal situation would be to be able to have a version of this test for ourselves, so we can run it before merging code (not to mention before an app release is cut by another team, sent to the store, and whatnot).

Easier said than done, of course; but that'd be our holy grail for testing this feature.

File: src/objective-c/tests/CronetUnitTests/CronetUnitTests.m [hunk 5]
```diff
@@ -270,8 +270,11 @@ - (void)testInternalError {
 }
 
 - (void)PacketCoalescing:(bool)use_coalescing {
```
Comment (jcanizales) [inline src/objective-c/tests/CronetUnitTests/CronetUnitTests.m]: nit:
* methods and parameters are named in lowerCamelCase
* s/bool/BOOL/

File: src/objective-c/tests/CronetUnitTests/CronetUnitTests.m [hunk 6]
```diff
@@ -270,8 +270,11 @@ - (void)testInternalError {
...
-  grpc_cronet_use_packet_coalescing(use_coalescing);
-
+  grpc_arg arg;
+  arg.key = GRPC_ARG_USE_CRONET_PACKET_COALESCING;
+  arg.type = GRPC_ARG_INTEGER;
+  arg.value.integer = use_coalescing ? 1:0;
+  grpc_channel_args *args = grpc_channel_args_copy_and_add(NULL, &arg, 1);
```
Comment (jcanizales) [inline src/objective-c/tests/CronetUnitTests/CronetUnitTests.m]: I'm pretty sure we have somewhere a constructor for `grpc_channel_args` from a `NSDictionary` that you can use to make this clear.

File: src/objective-c/tests/CronetUnitTests/CronetUnitTests.m [hunk 7]
```diff
@@ -436,8 +439,8 @@ - (void)PacketCoalescing:(bool)use_coalescing {
 }
 
 - (void)testPacketCoalescing {
-  [self PacketCoalescing:false];
   [self PacketCoalescing:true];
+  [self PacketCoalescing:false];
```
Comment (jcanizales) [inline src/objective-c/tests/CronetUnitTests/CronetUnitTests.m]: s/true, false/YES, NO/

File: src/objective-c/tests/CronetUnitTests/CronetUnitTests.m (hunk 6 above)
Comment (muxi) [inline src/objective-c/tests/CronetUnitTests/CronetUnitTests.m]: Since things being tested here are core changes, I would rather not to invoke something in ObjC library. Besides that I feel it is a bit over weight to move that code here and create `grpc_channel_args` with it; I would more like to stay with this core tests style of creating channel args.

File: src/objective-c/tests/CronetUnitTests/CronetUnitTests.m [hunk 8]
```diff
@@ -273,7 +273,7 @@ - (void)testInternalError {
   grpc_completion_queue_destroy(cq);
 }
 
-- (void)PacketCoalescing:(bool)use_coalescing {
+- (void)packetCoalescing:(BOOL)use_coalescing {
```
Comment (jcanizales) [inline src/objective-c/tests/CronetUnitTests/CronetUnitTests.m]: Parameter names too :)

Comment (muxi) [inline src/objective-c/tests/CronetUnitTests/CronetUnitTests.m]: That is right. Fixed :)

**Comment by muxi** [comment]
Ping @ctiller for changes in 0a2fae9 where 
1. Creation of cronet transport is moved to `cronet_transport.c` as we talked about, and
2. enabling/disabling of packet coalescing feature is done by channel args

**Comment by muxi** [comment]
test this please

**Comment by muxi** [comment]
Jenkins: test this please

**Comment by muxi** [comment]
Test this please

**Comment by muxi** [comment]
Failed tests:
gRPC_performance_pull_requests - #9668
